src/TesApi.Web/CachingAzureProxy.cs

```diff
@@ -114,18 +109,22 @@ public async Task<StorageAccountInfo> GetStorageAccountInfoAsync(string storageA
         }
 
         ///<inheritdoc/>
-        public async Task<List<VirtualMachineInfo>> GetVmSizesAndPricesAsync()
+        public Task<List<VirtualMachineInfo>> GetVmSizesAndPricesAsync()
         {
             const string key = "vmSizesAndPrices";
 
-            if (!cache.TryGetValue(key, out List<VirtualMachineInfo> cachedVmSizesAndPrices))
+            return cache.GetOrCreateAsync(key, cachedVmSizesAndPrices =>
             {
-                // retry
-                var vmSizesAndPrices = await azureProxy.GetVmSizesAndPricesAsync();
-                cache.Set(key, vmSizesAndPrices.ToList(), TimeSpan.FromDays(1));
-            }
+                cachedVmSizesAndPrices.AddExpirationToken(new CancellationChangeToken(new CancellationTokenSource(TimeSpan.FromDays(1)).Token));
+
+                cachedVmSizesAndPrices.RegisterPostEvictionCallback((key, value, reason, state) =>
```

Comment:
I think it would be better to run a separate thread in an infinite loop that updates the cache _before_ eviction - i.e. if the object is cached for 1 day, have a separate loop that runs every 23 hours and updates the cached object.

Consider this scenario:

1.  The system is very active (lots of new TesTasks being created)
2.  The object is evicted from the cache, spawning a callback thread.  The callback thread takes ~1 minute to complete
3.  In the mean time, (5) more TesTasks are run during that minute, and the object is not in the cache, so they each individually call GetVmSizesAndPricesAsync(), which results in all (5) TesTasks either being delayed and/or their requests failing and the logic falling back to the hard-coded pricing logic.

Reply:
LazyCache package (a thin wrapper around MemoryCache) solves this problem. 
In case where the item is not in cache and multiple clients hit it at the same time, only the first client will be allowed to the underlying expensive call. All other clients will wait for the first one to complete. Similarly, if the cache expires and is being refreshed by the callback thread, clients that come in at that moment will wait for the callback thread to finish and will use the returned (now cached) value without calling the underlying code.